euphonic/force_constants.py

```diff
@@ -445,19 +449,28 @@ def _calculate_phonons_at_qpts(
             reigenvecs = np.zeros(
                 (0, 3*n_atoms, n_atoms, 3), dtype=np.complex128)
 
+        euphonic_path = os.path.dirname(euphonic.__file__)
+        cext_err_msg = (f'Euphonic\'s C extension couldn\'t be imported '
+                        f'from {euphonic_path}, it may not have been '
+                        f'installed.')
         try:
-            if use_c:
+            if use_c is not False:
                 try:
                     import euphonic._euphonic as euphonic_c
                 except ImportError:
-                    warnings.warn((
-                        'use_c=True is set, but the Euphonic\'s C '
-                        'extension couldn\'t be imported, it may not '
-                        'have been installed. Attempting to fall back '
-                        'to pure Python calculation'), stacklevel=2)
+                    if use_c is None:
+                        warnings.warn((
+                            cext_err_msg
+                            + ' Falling back to pure Python calculation.'),
+                            stacklevel=3)
                     raise
             else:
                 raise ImportError

```

Comment:
I do find this stack of if functions and try/Except a bit tricky to follow. I'm also wary also of using try/except/raise as a form of control logic over a large block of code. What happens if something other than `import euphonic._euphonic` raises an ImportError? Then it shouldn't be excepted. I think this was motivated by the problem of reaching the python implementation from different paths. How about instead we put that in a function and have something like:

```python
if use_c is not False:
    try:
        import euphonic._euphonic as euphonic_c
    except ImportError:
        if use_c:
            raise
        else:
            warnings.warn(...)
            _do_it_with_python(*args)
    else:
        _do_it_with_c(*args)
else:
    _do_it_with_python(*args)
```

Reply:
Ok done! I've implemented it a little differently, as I found handling the different output arguments for Python implementation (depending on whether `return_eigenvectors=True/False`) was kind of awkward. But I think it works!